### ISSUE TITLE
Objectives for 2025.

### ISSUE THREAD

**Issue by righettod** [OWASP/www-project-secure-headers#201]
# 🔵TODO

- [ ] Handle #212
- [ ] Create a presentation of the project using the following guidelines:

![image](https://github.com/user-attachments/assets/dc75ecc3-e307-4b1d-a8c7-8e068e75eec8)

- [ ] Track https://github.com/akka/akka-http/issues/155 to add it to the **technical resources** section once it will be implemented.
- [ ] Track [Local Network Access](https://github.com/explainers-by-googlers/local-network-access) evolution and implementation to see if we can add it to the **miscellaneous** section.
- [ ] Add information for #261 

# 🟡Pending by the OSHP team

- [ ] Handle #211

# 🟠Pending by a third party

None

# 🟢Done

- [x] Try to implement #210 
- [x] Find new HTTP response headers disclosing technical information, this using GEMINI or ChatGPT via a specific crafted prompt.
- [x] Track this [RUST project](https://github.com/danielkov/rust-helmet) to add it to the **technical resources** section if we see that it became stable and maintained across the time.
- [x] See if it is consistent to add the header [X-DNS-Prefetch-Control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/X-DNS-Prefetch-Control).
- [x] Fix #251

# 🔴Cancelled

None

**Comment by righettod** [comment]
> 🚧 Create a presentation of the project.

[OWASPLightThemeTemplate.pptx](https://github.com/user-attachments/files/18181564/OWASPLightThemeTemplate.pptx)

**Comment by righettod** [comment]
> 🚧 Try to implement #210 .

See PR #217

**Comment by righettod** [comment]
✅ #210 done

**Comment by righettod** [comment]
> 🚧 Find new HTTP response headers disclosing technical information, this using GEMINI or ChatGPT via a specific crafted prompt.

# Input

Prompt used: `give me a list of HTTP response header the disclose technical information. exclude any header name present in the provided JSON file`.

I attached the file [headers_remove.json](https://raw.githubusercontent.com/OWASP/www-project-secure-headers/refs/heads/master/ci/headers_remove.json).

# Result from GEMINI

> Model `2.5 Flash`

Here are some HTTP response headers that can disclose technical information about the server or application:

* **X-Powered-By**: Indicates the technology (e.g., PHP, ASP.NET, Express) that generated the response.
* **Via**: Shows the intermediaries (proxies, gateways) the request passed through.
* **Server-Timing**: Provides performance metrics about the request-response cycle.
* **Alternate-Protocol**: Suggests alternative protocols the server supports.
* **X-AspNet-Version**: Specifies the version of ASP.NET used.
* **X-AspNetMvc-Version**: Indicates the version of ASP.NET MVC in use.
* **X-Generator**: Reveals the software or content management system used to generate the page.
* **X-Couchbase-SRS**: Can indicate the use of Couchbase Server.
* **X-Varnish**: Shows that Varnish Cache is being used and can include details like the Varnish instance ID.
* **X-Cache**: Indicates whether a response was served from a cache (e.g., HIT, MISS) and often includes the cache server's hostname.
* **X-Cache-Lookup**: Similar to X-Cache, providing more detailed cache lookup information.
* **X-UA-Compatible**: While primarily for rendering modes in Internet Explorer, it can sometimes hint at the server-side technology if it's dynamically set based on server logic.
* **X-Backend-Server**: Directly exposes the hostname or identifier of the backend server.
* **X-Served-By**: Similar to X-Backend-Server, indicating which server processed the request.
* **X-Runtime**: Often used in Ruby on Rails applications to show the processing time.
* **X-Version**: A generic header that might be used by applications to expose their version number.
* **ETag**: While not directly revealing technology, a predictable ETag generation pattern might indicate the underlying software.

Headers Excluded from the List

The following headers were excluded based on the `headers_remove.json` file:

* `$wsep`
* `Host-Header`
* `K-Proxy-Request`
* `Liferay-Portal`
* `OracleCommerceCloud-Version`
* `Pega-Host`
* `Powered-By`
* `Product`
* `Server`
* `SourceMap`
* `X-AspNet-Version`
* `X-AspNetMvc-Version`
* `X-Atmosphere-error`
* `X-Atmosphere-first-request`
* `X-Atmosphere-tracking-id`
* `X-B3-ParentSpanId`
* `X-B3-Sampled`
* `X-B3-SpanId`
* `X-B3-TraceId`
* `X-BEServer`
* `X-Backside-Transport`
* `X-CF-Powered-By`
* `X-CMS`
* `X-CalculatedBETarget`
* `X-Cocoon-Version`
* `X-Content-Encoded-By`
* `X-DiagInfo`
* `X-Envoy-Attempt-Count`
* `X-Envoy-External-Address`
* `X-Envoy-Internal`
* `X-Envoy-Original-Dst-Host`
* `X-Envoy-Upstream-Service-Time`
* `X-FEServer`
* `X-Framework`
* `X-Generated-By`
* `X-Generator`
* `X-Jitsi-Release`
* `X-Joomla-Version`
* `X-Kubernetes-PF-FlowSchema-UI`
* `X-Kubernetes-PF-PriorityLevel-UID`
* `X-LiteSpeed-Cache`
* `X-LiteSpeed-Purge`
* `X-LiteSpeed-Tag`
* `X-LiteSpeed-Vary`
* `X-Litespeed-Cache-Control`

Sources:

* https://medium.com/@zakeeandroid/version-disclosure-in-headers-and-response-security-threat-0b4e86272018
* https://www.upguard.com/blog/exposed-headers
* https://www.cachefly.com/news/the-role-and-importance-of-server-headers-in-http-communication/
* https://www.browserstack.com/guide/http-headers
* https://en.wikipedia.org/wiki/List_of_HTTP_header_fields
* (https://www.geeksforgeeks.org/web-tech/http-headers/
* https://m.youtube.com/watch?v=WTBt1cveEwo
* https://cqr.company/web-vulnerabilities/information-leakage-via-http-headers/

# Result from CHATGPT

> Model `ChatGPT`

Here’s a list of **HTTP response headers that are known to disclose technical information**, **excluding** those already present in your provided JSON file (`headers_remove.json`):

These headers may leak details about server technologies, frameworks, versions, or configurations:

1. `X-Powered-By-AspNet`
2. `X-Runtime`
3. `X-Version`
4. `X-Drupal-Cache`
5. `X-Drupal-Dynamic-Cache`
6. `X-Drupal-Cache-Contexts`
7. `X-Drupal-Cache-Tags`
8. `X-Drupal-Cache-Max-Age`
9. `X-CF-RAY`
10. `CF-Cache-Status`
11. `CF-RAY`
12. `CF-WAN-ID`
13. `X-Amz-Cf-Id`
14. `X-Amz-Id-2`
15. `X-Amz-Request-Id`
16. `X-Azure-Ref`
17. `X-MicrosoftAjax`
18. `X-Served-By`
19. `X-HW`
20. `X-Machine`
21. `X-Application-Context`
22. `Via`
23. `Backend`
24. `X-Instance-ID`
25. `X-Cluster-Node`
26. `X-Backend-Server`
27. `X-CDN`
28. `X-Capacity`
29. `X-Origin`
30. `X-Node`
31. `X-Server`
32. `X-Build-ID`
33. `X-Request-ID`
34. `X-Trace`
35. `X-Sentry-ID`
36. `X-Environment`
37. `X-Cloud-Trace-Context`

# Remark

🤔 I do not see interesting one to add.

@riramar Any remark on this?

**Comment by riramar** [comment]
For the task using GEMINI or ChatGPT, I don't have comments and I'd considered done with the data from your comment.
Now regarding the task "HTTP response headers disclosing technical information", not sure how to track that since there is no end.

**Comment by righettod** [comment]
> For the task using GEMINI or ChatGPT, I don't have comments and I'd considered done with the data from your comment. Now regarding the task "HTTP response headers disclosing technical information", not sure how to track that since there is no end.

Thanks, I will mark the task done and I will take a look at the header **X-Couchbase-Meta** to see if it is really interesting.

**Comment by righettod** [comment]
> 🚧 See if it is consistent to add the header [X-DNS-Prefetch-Control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/X-DNS-Prefetch-Control).

It is true that it is more related to a privacy aspect. Perhaps adding it is wrong from a consistency perspective for the project. I prefer to ask your point of view 😉

So what do you think @riramar about adding this header ? 🤔

Explanation of the header from ChatGTP combined with [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/X-DNS-Prefetch-Control):

The **`X-DNS-Prefetch-Control`** HTTP response header is used to **enable or disable DNS prefetching** in the browser.

**What DNS Prefetching Is**

* Modern browsers sometimes try to speed up page loading by **resolving domain names in advance**—before the user actually clicks a link.
* For example, if your page has a link to `https://example.com`, the browser might perform a DNS lookup for `example.com` while the user is still reading the current page.
* This means when the user finally clicks the link, the DNS resolution is already done, so the page loads faster.

**Security & Privacy Concerns**

While DNS prefetching can improve performance, it can also:

* **Leak information** about which domains a user might visit, even if they never click the link.
* Cause **unnecessary DNS traffic**, which may be undesirable in privacy-sensitive contexts.

**How `X-DNS-Prefetch-Control` Works**

This header tells the browser whether DNS prefetching should be allowed.

**Possible values:**

* `on` → Enable DNS prefetching.
* `off` → Disable DNS prefetching.

**Example to disable it:**

```http
X-DNS-Prefetch-Control: off
```

**Example to enable it:**

```http
X-DNS-Prefetch-Control: on
```

**When You Might Use It**

* **Disable** (`off`) for privacy-sensitive applications (banking sites, medical portals, admin panels).
* **Enable** (`on`) if you want to maximize performance and have no privacy concerns.

**Comment by righettod** [comment]
> 🚧  Track this [RUST project](https://github.com/danielkov/rust-helmet) to add it to the **technical resources** section if we see 
that it became stable and maintained across the time.

Added via PR #237
